client/consensus/aura/src/collators/lookahead.rs

```diff
@@ -118,183 +142,243 @@ pub async fn run<Block, P, BI, CIDP, Client, Backend, RClient, SO, Proposer, CS>
+				relay_parent,
+				para_id: params.para_id,
+				ancestry_lookback: max_ancestry_lookback(relay_parent, &params.relay_client).await,
+				max_depth: PARENT_SEARCH_DEPTH,
+				ignore_alternative_branches: true,
+			};
+
+			let potential_parents =
+				cumulus_client_consensus_common::find_potential_parents::<Block>(
+					parent_search_params,
+					&*params.para_backend,
+					&params.relay_client,
+				)
+				.await;
+
+			let mut potential_parents = match potential_parents {
+				Err(e) => {
+					tracing::error!(
+						target: crate::LOG_TARGET,
+						?relay_parent,
+						err = ?e,
+						"Could not fetch potential parents to build upon"
+					);
+
+					continue
 				},
 				Ok(x) => x,
 			};
 
-			match collator
-				.collate(
-					&parent_header,
-					&slot_claim,
-					None,
-					(parachain_inherent_data, other_inherent_data),
-					params.authoring_duration,
-					// Set the block limit to 50% of the maximum PoV size.
-					//
-					// TODO: If we got benchmarking that includes the proof size,
-					// we should be able to use the maximum pov size.
-					(validation_data.max_pov_size / 2) as usize,
+			let included_block = match potential_parents.iter().find(|x| x.depth == 0) {
+				None => continue, // also serves as an `is_empty` check.
+				Some(b) => b.hash,
+			};
+
+			let para_client = &*params.para_client;
+			let keystore = &params.keystore;
+			let can_build_upon = |block_hash| {
+				can_build_upon::<_, _, P>(
+					slot_now,
+					timestamp,
+					block_hash,
+					included_block,
+					para_client,
+					&keystore,
 				)
-				.await
-			{
-				Ok((_collation, block_data, new_block_hash)) => {
-					parent_hash = new_block_hash;
-					parent_header = block_data.into_header();
+			};
 
-					// Here we are assuming that the import logic protects against equivocations
-					// and provides sybil-resistance, as it should.
-					collator.collator_service().announce_block(new_block_hash, None);
+			// Sort by depth, ascending, to choose the longest chain.
+			//
+			// If the longest chain has space, build upon that. Otherwise, don't
+			// build at all.
+			potential_parents.sort_by_key(|a| a.depth);
+			let initial_parent = match potential_parents.pop() {
+				None => continue,
+				Some(p) => p,
+			};
 
-					// TODO [https://github.com/paritytech/polkadot/issues/5056]:
-					// announce collation to relay-chain validators.
-				},
-				Err(err) => {
-					tracing::error!(target: crate::LOG_TARGET, ?err);
-					break
-				},
+			// Build in a loop until not allowed. Note that the authorities can change
+			// at any block, so we need to re-claim our slot every time.
+			let mut parent_hash = initial_parent.hash;
+			let mut parent_header = initial_parent.header;
+			let overseer_handle = &mut params.overseer_handle;
+
+			// This needs to change to support elastic scaling, but for continuously
+			// scheduled chains this ensures that the backlog will grow steadily.
+			for n_built in 0..2 {
```

Comment:
I guess this should use `max_ancestry_lookback` which already hardcodes 2.

Reply:
This is actually more of a velocity hardcoding than ancestry lookback. It will need to be adjusted when we start dealing with dynamic velocities, but saving that for a later date as velocity won't exceed 1 per relay-block for a while.

The specific behavior we're dealing with is that most parachains hardcode an Aura slot duration of 12 seconds, so they can actually author at 2 different relay parents. These chains should set velocity=2 in their runtime as they expect 2 blocks to be included per slot. That allows them to author 3 blocks per slot when the unincluded segment has space, and this change just spreads that out and showed better results in testing.

Generalizing a bit - any slot duration which is a multiple of the relay chain slot duration probably doesn't want to try authoring all the blocks in their slot at the front, but rather spread them out.

It's definitely not ideal or particularly elegant, but couldn't think of a better solution at the time.

The alternative is to keep track in `aura-ext` how many blocks are authored per relay-chain slot not per parachain slot, and change `can_build_upon` to `can_build_upon(included_hash, relay_chain_slot) . But then slot durations become kind of meaningless in a relay-chain setting, except for rotating keys.

WDYT?

Reply:
Yes, my initial comment is not correct, first I was confused where does hardcoded 2 come from, then noticed the below function and assumed it's the same.

We had this discussion about filling backlog in the first PR, and I guess looping while `can_build` is ok for the start. When I was thinking about this, I imagined parachains would somehow customize this rate of block production, like
```rust
enum BlockProductionRate {
    /// Only author every N seconds.
    Fixed(Duration),
    /// No fixed block time, attempt to author whenever possible,
    Asap,
    // ... and more
}
```

Reply:
Let's revisit this in a new issue. The design space is pretty open, though I think the best the runtime can do is give "hints" and not expect them to be obeyed.